[FLINK-5311] Add user documentation for bipartite graph

mushketyk opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+---
+title: Graph Generators
```
This shouldn't be Graph Generators I believe :)

mushketyk replied:
Oops, right :)

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+Bipartite Graph
+---------------
+
+A bipartite graph (also called a two-mode graph) is a type of graph where vertices are separated into two disjoint sets. These sets are usually called top and bottom vertices. A single edge in this graph can only connect vertices from opposite sets (i.e. bottom vertex to top vertex) and cannot connect to vertices in the same set.
```
A single edge => an edge
cannot connect to vertices => cannot connect *two vertices

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+
+A bipartite graph (also called a two-mode graph) is a type of graph where vertices are separated into two disjoint sets. These sets are usually called top and bottom vertices. A single edge in this graph can only connect vertices from opposite sets (i.e. bottom vertex to top vertex) and cannot connect to vertices in the same set.
+
+Theses graphs have wide application in practice and can be a more natural choice for particular domains. For example to represent authorship of scientific papers top vertices can represent scientific papers while bottom nodes will represent authors. Naturally a node between a top and a bottom nodes would represent an authorship of a particular scientific paper. Another common example for applicat ...
```
a node between a top and a bottom nodes => an edge?

vasia replied:
a relationships => relationships

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+
+Theses graphs have wide application in practice and can be a more natural choice for particular domains. For example to represent authorship of scientific papers top vertices can represent scientific papers while bottom nodes will represent authors. Naturally a node between a top and a bottom nodes would represent an authorship of a particular scientific paper. Another common example for applicat ...
+
+Bipartite graph are used instead of regular graphs (one-mode) for the following practical [reasons](http://www.complexnetworks.fr/wp-content/uploads/2011/01/socnet07.pdf):
```
*graphs

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+Theses graphs have wide application in practice and can be a more natural choice for particular domains. For example to represent authorship of scientific papers top vertices can represent scientific papers while bottom nodes will represent authors. Naturally a node between a top and a bottom nodes would represent an authorship of a particular scientific paper. Another common example for applicat ...
+
+Bipartite graph are used instead of regular graphs (one-mode) for the following practical [reasons](http://www.complexnetworks.fr/wp-content/uploads/2011/01/socnet07.pdf):
+ * They preserve more information about a connection between vertices. For example instead of a single link between two researchers in a graph that represents that they authored a paper together a bipartite graph preserve the information about what papers they authored
```
*preserves

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+
+Bipartite graph are used instead of regular graphs (one-mode) for the following practical [reasons](http://www.complexnetworks.fr/wp-content/uploads/2011/01/socnet07.pdf):
+ * They preserve more information about a connection between vertices. For example instead of a single link between two researchers in a graph that represents that they authored a paper together a bipartite graph preserve the information about what papers they authored
+ * Bipartite graph can encode the same information more compactly than one-mode graphs
```
*graphs

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+--------------------
+
+A `BipartiteGraph` is represented by:
+ * `DataSet` of top nodes
```
*a DataSet...

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+ * `DataSet` of bottom nodes
+ * `DataSet` of edges between top and bottom nodes
+
+As in the `Graph` class nodes are represented by the `Vertex` type and the same rules applies to its types and values.
```
*apply

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+
+As in the `Graph` class nodes are represented by the `Vertex` type and the same rules applies to its types and values.
+
+The graph edges are represented by the `BipartiteEdge` type. An `BipartiteEdge` is defined by a top ID (the ID of the top `Vertex`), a bottom ID (the ID of the bottom `Vertex`) and an optional value. The main difference between the `Edge` and `BipartiteEdge` is that IDs of nodes it links can be of different types. Edges with no value have a `NullValue` value type.
```
*A BipartiteEdge

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+// TODO: Should be added when Scala interface is implemented
```
Don't leave a TODO in the docs. Either state explicitly that BipartiteGraph currently only exists in the Java API or we should make sure to implement the Scala methods before merging this.

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+// TODO: Should be added when Scala interface is implemented
```
same as above

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+---------------------
+
+
+* <strong>Projection</strong>: Projection is a common operation for bipartite graphs that converts a bipartite graph into a regular graph. There are two types of projections: top and bottom projections. Top projection preserves only top nodes in the result graph and create a link between them in a new graph only if there is an intermediate bottom node both top nodes connect to in the original gra ...
```
*creates

vasia replied:
*nodes

vasia replied:
Can you add a figure to illustrate a top and a bottom projection?

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+
+Gelly supports two sub-types of projections: simple projections and full projections. The only difference between them is what data is associated with edges in the result graph.
+
+In case of a simple projection each node in the result graph contains a pair of values of bipartite edges that connect nodes in the original graph:
```
*the case

vasia commented on docs/dev/libs/gelly/bipartite_graph.md:
```diff
@@ -0,0 +1,148 @@
+ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+Graph<String, String, Long, Long, Double> bipartiteGraph = ...;
+
+Graph<String, Long, Tuple2<Double, Double>> graph bipartiteGraph.projectionBottomSimple();
```
Again here, I believe we should either add a figure to illustrate or do something with the resulting projections in the code example to show why they are useful.

mushketyk commented:
Hi @vasia,

Thank you for your review. I've updated the documentation accordingly and added more detailed descriptions for projections code examples. Do you think it is still better to add some images?

vasia commented:
Thank you for the update @mushketyk! I still don't see any link from the Gelly guide page to the bipartite docs though. Can you please add that too? Otherwise people won't be able to find the docs :)
As for the images, I think it would be nice to have show how a projection works.

mushketyk commented:
Hi @vasia. Thank you for your comments. I've added the link to the doc and added the image (I tried to make it in the same style as other Gelly image).

vasia commented:
Hi @mushketyk, thank you for the update!
Just a couple of small things and we can merge:
- Can you add a note in the beginning of the docs that bipartite graphs are only currently supported in the Gelly Java API?
- I would rename the "Graph transformations" section to "Projection".

mushketyk commented:
Hi @vasia , thank you for your review!

I've added the warning to the Bipartite Graph page.
I am a bit reluctant to rename "Graph transformations" though since I wanted to extend this section in other commits when we have more operations for BipartiteGraph transformations, e.g. map.

What do you think about this?

vasia commented:
Thank you @mushketyk. That's OK. I'm merging this PR.
